### ISSUE TITLE
Composer Integration and project specific moduls?

### ISSUE THREAD

**Issue by tobi-pb** [magento/magento2#1342]
In our magento-1 project wie are using composer to install external dependencies (magento-core, germansetup, some extensions). Customer specific stuff is locacted in src/"Modulename".

This opens up possiblity to have composer.lock + composer.json and customer specfic extensions or themes in same git repository. 

As I unterstand current composer integration i have to create minimum 2 repositories for each magento customer? first  contains composer.json + composer.json ONLY, second contains all other customer specfic stuff. Do I have this right?  

In case I got it right: Could we implement a solution to use one repo?

**Comment by alankent** [comment]
We are not prescribing how people use Git.

One approach is to use composer to download external dependencies then Git commit everything (including external dependencies). You could call this the paranoid pattern.

Another is to gitignore external dependencies (there is work to keep them under vendor/ so this is a single directory).

You can certainly put local modules in same Git repo as composer.json file - just add the custom code dependencies to root composer. Json file

**Comment by tobi-pb** [comment]
Mh, I am not sure I have explained my problem exactly.

As an ageny we are supporting more than one shop and i like to have one GIT Repo with magento core packages. All shops are depending on this core repository. 

So I created a new project using

```
composer create-project magento/product-community-edition magento2 -s dev --prefer-dist
```

In this case i got a new project including a lot of php files which i NEVER NEVER allowed to change in my project:

```
  app/
   bin/
   ...
```

All this files are supplied from dependecy magento/product-community-edition, isn't it?
So i desided to add them to gitignore - excluding composer.json and composer.lock.

I provide that in my GIT Repo at https://github.com/tobi-pb/Magento2TestComposer

At next I tried to create a new clone of this repo.

```
git clone git@github.com:tobi-pb/Magento2TestComposer.git .
```

And installed:

```
composer install
```

Great job, everything works fine. 

So, may we could cleanup https://github.com/magento/magento2-community-edition a little bit?

I think we have to add following files to .gitignore because they exists in vendor as well as in root dir:

```
.htaccess
.htaccess.sample
.php_cs
.travis.yml
CHANGELOG.md
CONTRIBUTING.md
CONTRIBUTOR_LICENSE_AGREEMENT.html
COPYING.txt
Gruntfile.js
LICENSE.txt
LICENSE_AFL.txt
README.md
app/
bin/
dev/
index.php
lib/
nginx.conf.sample
package.json
php.ini.sample
pub/
setup/
update/
var/
```

Alternative we could add 

```
"extra":{
      "magento-root-dir": "mage"
    }
```

to composer.json add only add mage to .gitignore - sounds a little bit cleaner.

Thanks for reading ;)

Tobi

**Comment by Flyingmana** [comment]
In an agency, you would have only one git repositorie per customer, the composer.json you have would look something like this:

```json
{

    "repositories": [
        {
            "type": "composer",
            "url": "http://packages.magento.com"
        }
    ],
    "minimum-stability": "dev",
    "prefer-stable": true,
    "require": {
        "magento/product-community-edition": "0.74.0-beta12"
    }
}
```

!!!! ***caution, thats currently a lot of packages, and some big ones, so download can be slow.***!!

maybe a simple `/pub/` with a default index.php and stuff.

Your customer specific modules/code could then be placed in an extra directory.


note: this way to install is not finished yet and I assume will need some months, as supporting/documenting it, would make debugging core issues a lot harder as long as the core gets regular changes.


But thats only what I would expect, I have no idea about the real plans regarding this :)

**Comment by tobi-pb** [comment]
Thanks @Flyingmana, I allready got it, testet it and I find some improvements :)

Custom directory like "src" would be great but it is not supported at the moment (as you said) and it its not clear that your code would be merged to the official magento composer fork isn't it? So let us focus and discuss a simple solution together with magento. 

Would be great to use that in our first magento 2 projects :-)

Tobi

**Comment by tobi-pb** [comment]
As @Flyingmana told me "src"-problem is solved when #1206 is merged. So .gitignore default values are open here only :)

**Comment by slavvka** [comment]
Hi @tobi-pb. Thank you for your report. 
The issue has been fixed in magento/magento2#28220 by @gabrieldagama in 2.4.0-develop branch
Related commit(s):
 - [a99046895db46350951b5e0daa290f3dc189d525](https://github.com/magento/magento2/commit/a99046895db46350951b5e0daa290f3dc189d525)
 - [8f841a045f586e95308934dce32cbc1aeca3d621](https://github.com/magento/magento2/commit/8f841a045f586e95308934dce32cbc1aeca3d621)
 - [b1f56c4f8e95f63c28accb2db2726ab031291c8d](https://github.com/magento/magento2/commit/b1f56c4f8e95f63c28accb2db2726ab031291c8d)
 - [8eae1b540c0698a2e0c46a03087578da7c39e8f9](https://github.com/magento/magento2/commit/8eae1b540c0698a2e0c46a03087578da7c39e8f9)
 - [bfa589af67418e74ce5358caa5b47f1f78295c8e](https://github.com/magento/magento2/commit/bfa589af67418e74ce5358caa5b47f1f78295c8e)

The fix will be available with the upcoming 2.4.0 release.
